Firestore/core/src/firebase/firestore/remote/grpc_root_certificate_finder_apple.mm

```diff
@@ -34,28 +34,53 @@
 using util::StatusOr;
 using util::StringFormat;
 
-std::string LoadGrpcRootCertificate() {
-  // Try to load certificates bundled by gRPC-C++ if available (depends on
-  // gRPC-C++ version).
-  // Note that `mainBundle` may be nil in certain cases (e.g., unit tests).
-  NSBundle* bundle = [NSBundle bundleWithIdentifier:@"org.cocoapods.grpcpp"];
+NSString* FindPathToCertificatesFile() {
+  // Certificates file might be present in one of several bundles, based on
+  // the environment.
+  NSArray<NSBundle*>* bundles = @[
+    // First, try to load certificates bundled by gRPC-C++ if available
+    // (pod versions 0.0.6+).
+    [NSBundle bundleWithIdentifier:@"org.cocoapods.grpcpp"],
+    // Fall back to the certificates bundled with Firestore if necessary.
+    [NSBundle bundleForClass:FSTFirestoreClient.class],
+    // Finally, users manually adding resources to the project may add the
+    // certificate to the main application bundle. Note that `mainBundle` is nil
+    // for unit tests of library projects, so it cannot fully substitute for
+    // checking framework bundles.
+    [NSBundle mainBundle],
+  ];
+
   NSString* path;
-  if (bundle) {
+  int i = 0;
+  for (NSBundle* bundle in bundles) {
+    ++i;
+    LOG_DEBUG("roots.pem file: searching, iteration %s", i);
+    if (!bundle) {
+      continue;
+    }
+    LOG_DEBUG("roots.pem file: loaded bundle %s", [bundle bundleIdentifier]);
+
     path =
         [bundle pathForResource:@"gRPCCertificates.bundle/roots" ofType:@"pem"];
+    if (!path) {
+      // Technically, certificate with this name cannot be present in gRPC-C++
+      // framework, but checking for it unconditionally makes logic here
+      // simpler.
+      path = [bundle pathForResource:@"gRPCCertificates-Firestore.bundle/roots"
+                              ofType:@"pem"];
+    }
+    if (path) {
+      LOG_DEBUG("Using roots.pem file from bundle %s",
```

Comment:
This ends up not logging any confirmation of not-found a the path but could log upwards of 9 lines in the successful case. I'd suggest trying to shoot for at most one log line per iteration. Something like:

```objc
NSString* resource = @"gRPCCertificates.bundle/roots";
NSString* path = [bundle pathForResource:resource ofType:@"pem"];
if (!path) {
  resource = @"gRPCCertificates-Firestore.bundle/roots";
  path = [bundle pathForResource:resource ofType:@"pem"];
}

if (path) {
  LOG_DEBUG("%s.pem found in bundle %s", resource, [bundle bundleIdentifier]);
  return path;
} else {
  LOG_DEBUG("%s.pem not found in bundle %s", resource, [bundle bundleIdentifier]);
}
```

WDYT?

Reply:
Done. I like this better, thanks for the suggestion.